lib/compress/zstd_ldm.c

```diff
@@ -375,280 +420,151 @@ size_t ZSTD_compressBlock_ldm(
-                          offset + ZSTD_REP_MOVE, mLength - MINMATCH);
+            sequences->litLength = ip - anchor;
+            sequences->matchLength = mLength;
+            sequences->offset = offset;
+            ++sequences;
         }
 
         /* Insert the current entry into the hash table */
         ZSTD_ldm_makeEntryAndInsertByTag(ldmState, rollingHash, hBits,
                                          (U32)(lastHashed - base),
-                                         ldmParams);
+                                         *params);
 
         assert(ip + backwardMatchLength == lastHashed);
 
         /* Fill the hash table from lastHashed+1 to ip+mLength*/
         /* Heuristic: don't need to fill the entire table at end of block */
-        if (ip + mLength < ilimit) {
+        if (ip + mLength <= ilimit) {
             rollingHash = ZSTD_ldm_fillLdmHashTable(
                               ldmState, rollingHash, lastHashed,
-                              ip + mLength, base, hBits, ldmParams);
+                              ip + mLength, base, hBits, *params);
             lastHashed = ip + mLength - 1;
         }
         ip += mLength;
         anchor = ip;
-        /* Check immediate repcode */
-        while ( (ip < ilimit)
-             && ( (rep[1] > 0) && (rep[1] <= (U32)(ip-lowest))
-             && (MEM_read32(ip) == MEM_read32(ip - rep[1])) )) {
-
-            size_t const rLength = ZSTD_count(ip+4, ip+4-rep[1],
-                                              iend) + 4;
-            /* Swap repToConfirm[1] <=> repToConfirm[0] */
-            {
-                U32 const tmpOff = rep[1];
-                rep[1] = rep[0];
-                rep[0] = tmpOff;
-            }
-
-            ZSTD_storeSeq(seqStore, 0, anchor, 0, rLength-MINMATCH);
-
-            /* Fill the  hash table from lastHashed+1 to ip+rLength*/
-            if (ip + rLength < ilimit) {
-                rollingHash = ZSTD_ldm_fillLdmHashTable(
-                                ldmState, rollingHash, lastHashed,
-                                ip + rLength, base, hBits, ldmParams);
-                lastHashed = ip + rLength - 1;
-            }
-            ip += rLength;
-            anchor = ip;
-        }
     }
-
-    ZSTD_ldm_limitTableUpdate(ms, anchor);
-    ZSTD_ldm_fillFastTables(ms, cParams, anchor);
-
-    lastLiterals = blockCompressor(ms, seqStore, rep, cParams, anchor, iend - anchor);
-    ms->nextToUpdate = (U32)(iend - base);
-
-    /* Return the last literals size */
-    return lastLiterals;
+    /* Return the number of sequences generated */
+    return sequences - sequencesStart;
 }
 
-size_t ZSTD_compressBlock_ldm_extDict(
-        ldmState_t* ldmState, ZSTD_matchState_t* ms, seqStore_t* seqStore, U32 rep[ZSTD_REP_NUM],
-        ZSTD_CCtx_params const* params, void const* src, size_t srcSize)
+/**
+ * If the sequence length is longer than remaining then the sequence is split
+ * between this block and the next.
+ *
```

Comment:
Another strategy could be to reduce block size, ending where the "long match" starts ?
This avoids splitting a match between 2 blocks, and would likely compress a little bit better,
though if `matchSize` is larger `blockSize`, it will still be necessary to split it .

Reply:
That could be possible, though then that would have to be handled by whoever calls `ZSTD_compressBlock_internal()`, or we'll have to bubble up the information that we didn't actually compress the last few bytes through a few layers.

Since we have to handle the splitting when `matchSize > blockSize` anyways, I'd rather keep the complexity internal to the function.

We can look at it again once the code is actually needed, until then I'll `#if 0` it out.